subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java

```diff
@@ -38,15 +46,28 @@
     private final Convention _convention;
     private final IConventionAware _source;
     private final Set<String> _propertyNames;
+    private final Map<String, Boolean> _propertyNamesAsProviderMapping;
     private final Map<String, MappedPropertyImpl> _mappings = new HashMap<String, MappedPropertyImpl>();
 
     public ConventionAwareHelper(IConventionAware source, Convention convention) {
         this._source = source;
         this._convention = convention;
+        this._propertyNamesAsProviderMapping = ClassInspector.inspect(source.getClass()).getProperties().stream().collect(Collectors.toMap(it -> it.getName(), it -> isProviderType(it.getGetters())));
         this._propertyNames = JavaPropertyReflectionUtil.propertyNames(source);
```

Comment:
❓ Not sure if those two should be merged together.

Reply:
Definitely. I think this is going to kill performance, as we're reflecting on the whole type hierarchy for every instance created, without caching the result. Plus we're doing the same inspection again in class decoration (and again just after the task is created, if this happens to represent a task ). We shouldn't add another place where we visit the type hierarchy, so at the very least we should combine the property name and property type calculations here, then later we can look at how we merge the other places.